### PR TITLE
Support MSC3391: Account data deletion

### PR DESCRIPTION
**Pull request by kerryarchibald**
MSC: https://github.com/ShadowJonathan/matrix-doc/blob/account-data-delete/proposals/3391-account-data-delete.md

## Checklist

-   [ ] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Support MSC3391: Account data deletion ([\#2967](https://github.com/matrix-org/matrix-js-sdk/pull/2967)). Contributed by @kerryarchibald.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -3764,6 +3764,23 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+    public async deleteAccountData(eventType: string): Promise<void> {
+        const supportsMsc3391DeleteAccountData =
+            this.canSupport.get(Feature.AccountDataDeletion) !== ServerSupport.Unsupported;
+        // if deletion is not supported overwrite with empty content
+        if (!supportsMsc3391DeleteAccountData) {
+            await this.setAccountData(eventType, {});
+            return;
+        }
+        const path = utils.encodeUri("/user/$userId/account_data/$type", {
+            $userId: this.credentials.userId!,
```
Comment (weeman1337) [inline src/client.ts:3776]: This is a public method and can be called when no user is logged in. Means the `!` is probably incorrect here. We should handle the case when it is null.

File: src/client.ts [hunk 2]
```diff
@@ -3764,6 +3764,23 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+            $type: eventType,
+        });
+        return await this.http.authedRequest(Method.Delete, path, undefined, undefined, {
```
Comment (weeman1337) [inline src/client.ts:3779]: Not sure about the unstable process: But shouldn't we here not call the endpoint depending on the `this.canSupport.get(Feature.AccountDataDeletion)` stable/unstable value?

**Review by weeman1337 — COMMENTED** [review]
Some details questions. Looks good otherwise :+1:

File: src/client.ts (hunk 1 above)
Comment (kerryarchibald) [reply src/client.ts]: True, I just copied this part from the `setAccountData` method, which I assume relies on `authedRequest` to fail nicely when there is no logged in user. Added `getSafeUserId` to use here

**Review by weeman1337 — APPROVED** [review]
Looks good :rocket:

**Pull request merged**